tests/php/Dev/DeprecationTest.php

```diff
@@ -2,16 +2,40 @@
 
 namespace SilverStripe\Dev\Tests;
 
+use PHPUnit\Framework\Error\Deprecated;
+use SilverStripe\Core\Config\Config;
 use SilverStripe\Dev\Deprecation;
 use SilverStripe\Dev\SapphireTest;
-use SilverStripe\Dev\Tests\DeprecationTest\TestDeprecation;
+use SilverStripe\Dev\Tests\DeprecationTest\DeprecationTestObject;
 
 class DeprecationTest extends SapphireTest
 {
+    protected static $extra_dataobjects = [
+        DeprecationTestObject::class,
+    ];
+
+    private $oldHandler;
+
+    protected function setup(): void
+    {
+        // Use custom error handler for two reasons:
+        // - Filter out errors for deprecated class properities unrelated to this unit test
+        // - Allow the use of expectDeprecation(), which doesn't work with E_USER_DEPRECATION by default
+        //   https://github.com/laminas/laminas-di/pull/30#issuecomment-927585210
+        parent::setup();
+        $this->oldHandler = set_error_handler(function ($errno, $errstr) {
+            if (str_contains($errstr, 'SilverStripe\\Dev\\Tests\\DeprecationTest')) {
+                throw new Deprecated($errstr, $errno, '', 1);
+            } else {
+                // noop
+            }
+        }, E_USER_DEPRECATED);
```

Comment:
```suggestion
        $this->oldHandler = set_error_handler(function (int $errno, string $errstr, string $errfile, int $errline, array $errcontext) {
            if (str_contains($errstr, 'SilverStripe\\Dev\\Tests\\DeprecationTest')) {
                throw new Deprecated($errstr, $errno, '', 1);
            }
            $old = $this->oldHandler;
            return $old($errno, $errstr, $errfile, $errline, $errcontext);
        });
```
Currently this will only catch deprecation notices, but it _won't_ pass everything else back to the existing error handler.
It is more correct to [ass everything that we're not explicitly interested in handling back to the original error handler for processing.

Without doing this, anything that throws user warnings won't actually fail the tests like it normally would.

Reply:
Are you sure? The 2nd param of set_error_handler E_USER_DEPRECATED should filter it so that only errors of that type are 'caught' by the custom error handler:

e.g.

```
php -r '
  set_error_handler(function ($errno, $errstr) { echo "I caught the warning"; }, E_USER_DEPRECATED);
  user_error("I did not catch the warning", E_USER_WARNING);
'
PHP Warning:  I did not catch the warning in Command line code on line 3
```

```
php -r '
  set_error_handler(function ($errno, $errstr) { echo "I caught the warning"; },  E_USER_WARNING);
  user_error("I did not catch the warning", E_USER_WARNING);
'
I caught the warning
```

Reply:
PHP only has one active error handler at a time. Setting the 2nd parameter only changes what this error handler catches - but it still replaces the old handler.

I tested this by adding a `user_error('arbitrary notice', E_USER_NOTICE);` in a new test that has `$this->expectError()` and a normal `E_USER_NOTICE` was outputted without converting to an exception, so the test failed because it didn't get the exception it expected.

Reply:
Ah OK, makes sense, that filtering param is pretty bad in that case

I've updated code, though modified your recommended code a bit
- removed the 5th param because it's been removed in php8 - https://www.php.net/manual/en/function.set-error-handler.php
- added a check to see that $this->oldHandler `is_callable()`